Feature/runtime raw socket disable

morantr opened the pull request:
If raw socket fails to be opened due to insufficient permissions continue without it.

tomereli commented on agent/src/beerocks/monitor/monitor_rssi.cpp:
```diff
@@ -38,8 +38,6 @@ void monitor_rssi::stop()
         close(arp_socket);
         arp_socket = -1;
     }
-    mon_db       = nullptr;
```
Why did you delete these lines? It is not clear from the commit message.

morantr replied:
nullptr assignment is not needed and not contributing anything. When the monitor calls to `mon_rssi.start()` and fails, it calls to `stop()` BUT it continue calling to `process()` on each work cycle, which uses these pointers.

tomereli commented on agent/src/beerocks/monitor/monitor_rssi.cpp:
```diff
@@ -241,7 +247,7 @@ void monitor_rssi::process()
                 }
             }
-            if (!conf_disable_initiative_arp) {
+            if (arp_enabled() || !conf_disable_initiative_arp) {
```
Shouldn't this be `if (arp_enabled() && !conf_disable_initiative_arp)`?
And coming to think about it, shouldn't arp_enabled() be enough? i.e. move the conf_disable_initiative_arp check to inside the arp_enabled() API.

morantr replied:
Yes, you right it should be `&&`. But no, it can't be inside arp_enabled() since it is two different things.

tomereli commented on agent/src/beerocks/monitor/monitor_rssi.h:
```diff
@@ -43,6 +43,8 @@ class monitor_rssi {
     void send_rssi_measurement_response(std::string &sta_mac, monitor_sta_node *sta_node);
     void monitor_idle_station(std::string &sta_mac, monitor_sta_node *sta_node);
 
+    bool arp_enabled() { return arp_socket > 0; }
```
Here also add the initiative_arp config check.

morantr replied:
No, the inactive arp is only for inactive arp, and not related to general arp.

tomereli commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -273,10 +273,10 @@ void monitor_thread::after_select(bool timeout)
-                return;
+                // If monitor rssi failed to start, continue without it. It might failed due to
+                // insufficient permissions. Detailed error message is printed inside.
+                LOG(WARNING) << "mon_rssi.start() failed, ignore and continue without it";
```
We should log the failure reason - i.e. if its because of missing permissions or some other failure. This has to be perfectly clear from this log.

morantr replied:
We don't need it here. A detailed error will be printed inside (as I wrote on the comment here).

tomereli replied:
OK

tomereli commented on agent/src/beerocks/slave/backhaul_manager/wan_monitor.cpp:
```diff
@@ -33,7 +33,7 @@ static int netlink_open_socket()
     int fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
     if (fd < 0) {
-        LOG(ERROR) << "Failed creating netlink socket";
+        LOG(ERROR) << "Failed creating netlink socket: " << strerror(errno);
```
Good - will this print that there are no permissions to create the socket?

morantr replied:
yes

tomereli commented on framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp:
```diff
@@ -217,21 +219,21 @@ bool arp_monitor::start(std::string strIface)
     if ((m_fdMonSocket = socket(AF_NETLINK, SOCK_RAW | SOCK_CLOEXEC, NETLINK_ROUTE)) < 0) {
         LOG(ERROR) << "Failed opening socket: " << strerror(errno);
-        return false;
+        return -int(eErrorCode::OPERATION_NOT_SUPPORTED);
```
Why is the cast needed here?

morantr replied:
because the function returns `int`. 😕

tomereli commented on framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp:
```diff
@@ -217,21 +219,21 @@ bool arp_monitor::start(std::string strIface)
         LOG(ERROR) << "Failed setting SO_SNDBUF option: " << strerror(errno);
         stop();
-        return false;
+        return -1;
```
Why not return the enum as you do in other places?

morantr replied:
`-1` represents general failure without any specific error code.  I used the enum to separate permissions error form other general errors. Basically bpl_err.h doesn't have a general error enum, and I wasn't sure if it is a good idea to add one. Anyway `1` is already taken for `UNKNOWN` on bpl_err.h, unless you would like to replace the `-1` with `-int(eErrorCode::UNKNOWN)`.

tomereli reviewed (changes requested):
First commit message typo:
opening a raw a raw socket will fail. --> opening raw socket will fail.

tomereli reviewed (approved):
You answered all my comments so approving already.

morantr committed:
monitor_rssi: Continue working although failed to start

It is possible that due to insufficient permissions opening a raw
socket will fail.
The sending arp ability on monitor_rssi is not mandatory for system
work, and it can handle without it. In case that the  monitor_rssi fail
to open a raw socket for arps, and  the monitor will continue to work
without it, and use the process() function only for local measurements
that do not requirers arp.

In that case, the sta state will be always overridden to "IDLE".

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backhaul_manager: Skip eth link establishment if wan monitor fails

It is possible that due to insufficient permissions opening a raw socket
will fail.
Connecting with eth is necessary but not mandatory for the repeater to
connect backward, it can use instead wireless connection.
In case that the backhaul_manager fail to initialize the wan_monitor due
to failure to open a raw socket or any other reason, it will skip eth
backhaul connection establishment and skip to wireless backhaul link
establishment.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
arp_monitor:  Start function return int

Make te arp start function to return int instead of boolean, and return
"OPERATION_NOT_SUPPORTED" if opening raw socket fails.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
